File: Go/day06/day-06-a.go
package main

import (
	"fmt"
	"os"
	"strings"
)

func main() {
	bytes, _ := os.ReadFile("./input.txt")
	input := string(bytes)

	findFirstStartOfPacketMarker(input)
}


func findFirstStartOfPacketMarker(dataStream string) {
	chunks := windows(dataStream, 4)

	for i, chunk := range chunks {
		if isAllUnique(chunk) {
			fmt.Println(i +4)
			break
		}
	}
}

func windows(input string, windowSize int ) []string {
	chunks := []string{}

	for i := 0 ; i<= len(input) - windowSize; i++ {
		chunks = append(chunks, input[i: i+4])
	}

	return chunks
}

func isAllUnique(input string) bool {
	uniqueChars := ""

	for _, ch := range input {
		if !strings.ContainsRune(uniqueChars, ch) {
			uniqueChars += string(ch)
		}
	}

	return len(uniqueChars) == 4
}

File: Javascript/day06/day-06-a.js
const fs = require("fs");

const input = fs.readFileSync("./input.txt", "utf8");

findFirstStartOfPacketMarker(input);

function findFirstStartOfPacketMarker(dataStream) {
  let chunks = windows(dataStream, 4);

  for (let i = 0; i < chunks.length; i++) {
    const chunk = chunks[i];
    if (isAllUnique(chunk)) {
      console.log(i + 4);
      break;
    }
  }
}

/**
 *
 * @param {string} arr
 * @param {number} windowSize
 * @returns {string[]}
 */
function windows(input, windowSize) {
  let result = [];

  for (let i = 0; i <= input.length - windowSize; i++) {
    result.push(input.slice(i, i + windowSize));
  }

  return result;
}

/**
 *
 * @param {string} chunk
 * @returns {boolean}
 */
function isAllUnique(chunk) {
  let uniqueChars = [];

  for (let ch of chunk) {
    if (!uniqueChars.includes(ch)) {
      uniqueChars.push(ch);
    }
  }

  return uniqueChars.length == 4;
}
